// @aws-sdk/client-sqs/dist-cjs/index.js
'use strict';

var client$1 = require('@aws-sdk/core/client');
var middlewareSdkSqs = require('@aws-sdk/middleware-sdk-sqs');
var core = require('@smithy/core');
var client = require('@smithy/core/client');
var config = require('@smithy/core/config');
var endpoints = require('@smithy/core/endpoints');
var protocols = require('@smithy/core/protocols');
var retry = require('@smithy/core/retry');
var schema = require('@smithy/core/schema');
var httpAuthSchemeProvider = require('./auth/httpAuthSchemeProvider');
var runtimeConfig = require('./runtimeConfig');
var schemas_0 = require('./schemas/schemas_0');
var errors = require('./models/errors');
var SQSServiceException = require('./models/SQSServiceException');

const resolveClientEndpointParameters = (options) => {
    return Object.assign(options, {
        useDualstackEndpoint: options.useDualstackEndpoint ?? false,
        useFipsEndpoint: options.useFipsEndpoint ?? false,
        defaultSigningName: "sqs",
    });
};
const commonParams = {
    UseFIPS: { type: "builtInParams", name: "useFipsEndpoint" },
    Endpoint: { type: "builtInParams", name: "endpoint" },
    Region: { type: "builtInParams", name: "region" },
    UseDualStack: { type: "builtInParams", name: "useDualstackEndpoint" },
};

const getHttpAuthExtensionConfiguration = (runtimeConfig) => {
    const _httpAuthSchemes = runtimeConfig.httpAuthSchemes;
    let _httpAuthSchemeProvider = runtimeConfig.httpAuthSchemeProvider;
    let _credentials = runtimeConfig.credentials;
    return {
        setHttpAuthScheme(httpAuthScheme) {
            const index = _httpAuthSchemes.findIndex((scheme) => scheme.schemeId === httpAuthScheme.schemeId);
            if (index === -1) {
                _httpAuthSchemes.push(httpAuthScheme);
            }
            else {
                _httpAuthSchemes.splice(index, 1, httpAuthScheme);
            }
        },
        httpAuthSchemes() {
            return _httpAuthSchemes;
        },
        setHttpAuthSchemeProvider(httpAuthSchemeProvider) {
            _httpAuthSchemeProvider = httpAuthSchemeProvider;
        },
        httpAuthSchemeProvider() {
            return _httpAuthSchemeProvider;
        },
        setCredentials(credentials) {
            _credentials = credentials;
        },
        credentials() {
            return _credentials;
        },
    };
};
const resolveHttpAuthRuntimeConfig = (config) => {
    return {
        httpAuthSchemes: config.httpAuthSchemes(),
        httpAuthSchemeProvider: config.httpAuthSchemeProvider(),
        credentials: config.credentials(),
    };
};

const resolveRuntimeExtensions = (runtimeConfig, extensions) => {
    const extensionConfiguration = Object.assign(client$1.getAwsRegionExtensionConfiguration(runtimeConfig), client.getDefaultExtensionConfiguration(runtimeConfig), protocols.getHttpHandlerExtensionConfiguration(runtimeConfig), getHttpAuthExtensionConfiguration(runtimeConfig));
    extensions.forEach((extension) => extension.configure(extensionConfiguration));
    return Object.assign(runtimeConfig, client$1.resolveAwsRegionExtensionConfiguration(extensionConfiguration), client.resolveDefaultRuntimeConfig(extensionConfiguration), protocols.resolveHttpHandlerRuntimeConfig(extensionConfiguration), resolveHttpAuthRuntimeConfig(extensionConfiguration));
};

class SQSClient extends client.Client {
    config;
    constructor(...[configuration]) {
        const _config_0 = runtimeConfig.getRuntimeConfig(configuration || {});
        super(_config_0);
        this.initConfig = _config_0;
        const _config_1 = resolveClientEndpointParameters(_config_0);
        const _config_2 = client$1.resolveUserAgentConfig(_config_1);
        const _config_3 = retry.resolveRetryConfig(_config_2);
        const _config_4 = config.resolveRegionConfig(_config_3);
        const _config_5 = client$1.resolveHostHeaderConfig(_config_4);
        const _config_6 = endpoints.resolveEndpointConfig(_config_5);
        const _config_7 = middlewareSdkSqs.resolveQueueUrlConfig(_config_6);
        const _config_8 = httpAuthSchemeProvider.resolveHttpAuthSchemeConfig(_config_7);
        const _config_9 = resolveRuntimeExtensions(_config_8, configuration?.extensions || []);
        this.config = _config_9;
        this.middlewareStack.use(schema.getSchemaSerdePlugin(this.config));
        this.middlewareStack.use(client$1.getUserAgentPlugin(this.config));
        this.middlewareStack.use(retry.getRetryPlugin(this.config));
        this.middlewareStack.use(protocols.getContentLengthPlugin(this.config));
        this.middlewareStack.use(client$1.getHostHeaderPlugin(this.config));
        this.middlewareStack.use(client$1.getLoggerPlugin(this.config));
        this.middlewareStack.use(client$1.getRecursionDetectionPlugin(this.config));
        this.middlewareStack.use(middlewareSdkSqs.getQueueUrlPlugin(this.config));
        this.middlewareStack.use(core.getHttpAuthSchemeEndpointRuleSetPlugin(this.config, {
            httpAuthSchemeParametersProvider: httpAuthSchemeProvider.defaultSQSHttpAuthSchemeParametersProvider,
            identityProviderConfigProvider: async (config) => new core.DefaultIdentityProviderConfig({
                "aws.auth#sigv4": config.credentials,
            }),
        }));
        this.middlewareStack.use(core.getHttpSigningPlugin(this.config));
    }
    destroy() {
        super.destroy();
    }
}

class AddPermissionCommand extends client.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [endpoints.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("AmazonSQS", "AddPermission", {})
    .n("SQSClient", "AddPermissionCommand")
    .sc(schemas_0.AddPermission$)
    .build() {
}

class CancelMessageMoveTaskCommand extends client.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [endpoints.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("AmazonSQS", "CancelMessageMoveTask", {})
    .n("SQSClient", "CancelMessageMoveTaskCommand")
    .sc(schemas_0.CancelMessageMoveTask$)
    .build() {
}

class ChangeMessageVisibilityBatchCommand extends client.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [endpoints.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("AmazonSQS", "ChangeMessageVisibilityBatch", {})
    .n("SQSClient", "ChangeMessageVisibilityBatchCommand")
    .sc(schemas_0.ChangeMessageVisibilityBatch$)
    .build() {
}

class ChangeMessageVisibilityCommand extends client.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [endpoints.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("AmazonSQS", "ChangeMessageVisibility", {})
    .n("SQSClient", "ChangeMessageVisibilityCommand")
    .sc(schemas_0.ChangeMessageVisibility$)
    .build() {
}

class CreateQueueCommand extends client.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [endpoints.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("AmazonSQS", "CreateQueue", {})
    .n("SQSClient", "CreateQueueCommand")
    .sc(schemas_0.CreateQueue$)
    .build() {
}

class DeleteMessageBatchCommand extends client.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [endpoints.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("AmazonSQS", "DeleteMessageBatch", {})
    .n("SQSClient", "DeleteMessageBatchCommand")
    .sc(schemas_0.DeleteMessageBatch$)
    .build() {
}

class DeleteMessageCommand extends client.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [endpoints.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("AmazonSQS", "DeleteMessage", {})
    .n("SQSClient", "DeleteMessageCommand")
    .sc(schemas_0.DeleteMessage$)
    .build() {
}

class DeleteQueueCommand extends client.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [endpoints.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("AmazonSQS", "DeleteQueue", {})
    .n("SQSClient", "DeleteQueueCommand")
    .sc(schemas_0.DeleteQueue$)
    .build() {
}

class GetQueueAttributesCommand extends client.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [endpoints.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("AmazonSQS", "GetQueueAttributes", {})
    .n("SQSClient", "GetQueueAttributesCommand")
    .sc(schemas_0.GetQueueAttributes$)
    .build() {
}

class GetQueueUrlCommand extends client.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [endpoints.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("AmazonSQS", "GetQueueUrl", {})
    .n("SQSClient", "GetQueueUrlCommand")
    .sc(schemas_0.GetQueueUrl$)
    .build() {
}

class ListDeadLetterSourceQueuesCommand extends client.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [endpoints.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("AmazonSQS", "ListDeadLetterSourceQueues", {})
    .n("SQSClient", "ListDeadLetterSourceQueuesCommand")
    .sc(schemas_0.ListDeadLetterSourceQueues$)
    .build() {
}

class ListMessageMoveTasksCommand extends client.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [endpoints.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("AmazonSQS", "ListMessageMoveTasks", {})
    .n("SQSClient", "ListMessageMoveTasksCommand")
    .sc(schemas_0.ListMessageMoveTasks$)
    .build() {
}

class ListQueuesCommand extends client.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [endpoints.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("AmazonSQS", "ListQueues", {})
    .n("SQSClient", "ListQueuesCommand")
    .sc(schemas_0.ListQueues$)
    .build() {
}

class ListQueueTagsCommand extends client.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [endpoints.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("AmazonSQS", "ListQueueTags", {})
    .n("SQSClient", "ListQueueTagsCommand")
    .sc(schemas_0.ListQueueTags$)
    .build() {
}

class PurgeQueueCommand extends client.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [endpoints.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("AmazonSQS", "PurgeQueue", {})
    .n("SQSClient", "PurgeQueueCommand")
    .sc(schemas_0.PurgeQueue$)
    .build() {
}

class ReceiveMessageCommand extends client.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [
        endpoints.getEndpointPlugin(config, Command.getEndpointParameterInstructions()),
        client$1.getLongPollPlugin(config),
        middlewareSdkSqs.getReceiveMessagePlugin(config),
    ];
})
    .s("AmazonSQS", "ReceiveMessage", {})
    .n("SQSClient", "ReceiveMessageCommand")
    .sc(schemas_0.ReceiveMessage$)
    .build() {
}

class RemovePermissionCommand extends client.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [endpoints.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("AmazonSQS", "RemovePermission", {})
    .n("SQSClient", "RemovePermissionCommand")
    .sc(schemas_0.RemovePermission$)
    .build() {
}

class SendMessageBatchCommand extends client.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [
        endpoints.getEndpointPlugin(config, Command.getEndpointParameterInstructions()),
        middlewareSdkSqs.getSendMessageBatchPlugin(config),
    ];
})
    .s("AmazonSQS", "SendMessageBatch", {})
    .n("SQSClient", "SendMessageBatchCommand")
    .sc(schemas_0.SendMessageBatch$)
    .build() {
}

class SendMessageCommand extends client.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [
        endpoints.getEndpointPlugin(config, Command.getEndpointParameterInstructions()),
        middlewareSdkSqs.getSendMessagePlugin(config),
    ];
})
    .s("AmazonSQS", "SendMessage", {})
    .n("SQSClient", "SendMessageCommand")
    .sc(schemas_0.SendMessage$)
    .build() {
}

class SetQueueAttributesCommand extends client.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [endpoints.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("AmazonSQS", "SetQueueAttributes", {})
    .n("SQSClient", "SetQueueAttributesCommand")
    .sc(schemas_0.SetQueueAttributes$)
    .build() {
}

class StartMessageMoveTaskCommand extends client.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [endpoints.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("AmazonSQS", "StartMessageMoveTask", {})
    .n("SQSClient", "StartMessageMoveTaskCommand")
    .sc(schemas_0.StartMessageMoveTask$)
    .build() {
}

class TagQueueCommand extends client.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [endpoints.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("AmazonSQS", "TagQueue", {})
    .n("SQSClient", "TagQueueCommand")
    .sc(schemas_0.TagQueue$)
    .build() {
}

class UntagQueueCommand extends client.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [endpoints.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("AmazonSQS", "UntagQueue", {})
    .n("SQSClient", "UntagQueueCommand")
    .sc(schemas_0.UntagQueue$)
    .build() {
}

const paginateListDeadLetterSourceQueues = core.createPaginator(SQSClient, ListDeadLetterSourceQueuesCommand, "NextToken", "NextToken", "MaxResults");

const paginateListQueues = core.createPaginator(SQSClient, ListQueuesCommand, "NextToken", "NextToken", "MaxResults");

const commands = {
    AddPermissionCommand,
    CancelMessageMoveTaskCommand,
    ChangeMessageVisibilityCommand,
    ChangeMessageVisibilityBatchCommand,
    CreateQueueCommand,
    DeleteMessageCommand,
    DeleteMessageBatchCommand,
    DeleteQueueCommand,
    GetQueueAttributesCommand,
    GetQueueUrlCommand,
    ListDeadLetterSourceQueuesCommand,
    ListMessageMoveTasksCommand,
    ListQueuesCommand,
    ListQueueTagsCommand,
    PurgeQueueCommand,
    ReceiveMessageCommand,
    RemovePermissionCommand,
    SendMessageCommand,
    SendMessageBatchCommand,
    SetQueueAttributesCommand,
    StartMessageMoveTaskCommand,
    TagQueueCommand,
    UntagQueueCommand,
};
const paginators = {
    paginateListDeadLetterSourceQueues,
    paginateListQueues,
};
class SQS extends SQSClient {
}
client.createAggregatedClient(commands, SQS, { paginators });

const QueueAttributeName = {
    All: "All",
    ApproximateNumberOfMessages: "ApproximateNumberOfMessages",
    ApproximateNumberOfMessagesDelayed: "ApproximateNumberOfMessagesDelayed",
    ApproximateNumberOfMessagesNotVisible: "ApproximateNumberOfMessagesNotVisible",
    ContentBasedDeduplication: "ContentBasedDeduplication",
    CreatedTimestamp: "CreatedTimestamp",
    DeduplicationScope: "DeduplicationScope",
    DelaySeconds: "DelaySeconds",
    FifoQueue: "FifoQueue",
    FifoThroughputLimit: "FifoThroughputLimit",
    KmsDataKeyReusePeriodSeconds: "KmsDataKeyReusePeriodSeconds",
    KmsMasterKeyId: "KmsMasterKeyId",
    LastModifiedTimestamp: "LastModifiedTimestamp",
    MaximumMessageSize: "MaximumMessageSize",
    MessageRetentionPeriod: "MessageRetentionPeriod",
    Policy: "Policy",
    QueueArn: "QueueArn",
    ReceiveMessageWaitTimeSeconds: "ReceiveMessageWaitTimeSeconds",
    RedriveAllowPolicy: "RedriveAllowPolicy",
    RedrivePolicy: "RedrivePolicy",
    SqsManagedSseEnabled: "SqsManagedSseEnabled",
    VisibilityTimeout: "VisibilityTimeout",
};
const MessageSystemAttributeName = {
    AWSTraceHeader: "AWSTraceHeader",
    All: "All",
    ApproximateFirstReceiveTimestamp: "ApproximateFirstReceiveTimestamp",
    ApproximateReceiveCount: "ApproximateReceiveCount",
    DeadLetterQueueSourceArn: "DeadLetterQueueSourceArn",
    MessageDeduplicationId: "MessageDeduplicationId",
    MessageGroupId: "MessageGroupId",
    SenderId: "SenderId",
    SentTimestamp: "SentTimestamp",
    SequenceNumber: "SequenceNumber",
};
const MessageSystemAttributeNameForSends = {
    AWSTraceHeader: "AWSTraceHeader",
};

exports.$Command = client.Command;
exports.__Client = client.Client;
exports.SQSServiceException = SQSServiceException.SQSServiceException;
exports.AddPermissionCommand = AddPermissionCommand;
exports.CancelMessageMoveTaskCommand = CancelMessageMoveTaskCommand;
exports.ChangeMessageVisibilityBatchCommand = ChangeMessageVisibilityBatchCommand;
exports.ChangeMessageVisibilityCommand = ChangeMessageVisibilityCommand;
exports.CreateQueueCommand = CreateQueueCommand;
exports.DeleteMessageBatchCommand = DeleteMessageBatchCommand;
exports.DeleteMessageCommand = DeleteMessageCommand;
exports.DeleteQueueCommand = DeleteQueueCommand;
exports.GetQueueAttributesCommand = GetQueueAttributesCommand;
exports.GetQueueUrlCommand = GetQueueUrlCommand;
exports.ListDeadLetterSourceQueuesCommand = ListDeadLetterSourceQueuesCommand;
exports.ListMessageMoveTasksCommand = ListMessageMoveTasksCommand;
exports.ListQueueTagsCommand = ListQueueTagsCommand;
exports.ListQueuesCommand = ListQueuesCommand;
exports.MessageSystemAttributeName = MessageSystemAttributeName;
exports.MessageSystemAttributeNameForSends = MessageSystemAttributeNameForSends;
exports.PurgeQueueCommand = PurgeQueueCommand;
exports.QueueAttributeName = QueueAttributeName;
exports.ReceiveMessageCommand = ReceiveMessageCommand;
exports.RemovePermissionCommand = RemovePermissionCommand;
exports.SQS = SQS;
exports.SQSClient = SQSClient;
exports.SendMessageBatchCommand = SendMessageBatchCommand;
exports.SendMessageCommand = SendMessageCommand;
exports.SetQueueAttributesCommand = SetQueueAttributesCommand;
exports.StartMessageMoveTaskCommand = StartMessageMoveTaskCommand;
exports.TagQueueCommand = TagQueueCommand;
exports.UntagQueueCommand = UntagQueueCommand;
exports.paginateListDeadLetterSourceQueues = paginateListDeadLetterSourceQueues;
exports.paginateListQueues = paginateListQueues;
Object.prototype.hasOwnProperty.call(schemas_0, '__proto__') &&
    !Object.prototype.hasOwnProperty.call(exports, '__proto__') &&
    Object.defineProperty(exports, '__proto__', {
        enumerable: true,
        value: schemas_0['__proto__']
    });

Object.keys(schemas_0).forEach(function (k) {
    if (k !== 'default' && !Object.prototype.hasOwnProperty.call(exports, k)) exports[k] = schemas_0[k];
});
Object.prototype.hasOwnProperty.call(errors, '__proto__') &&
    !Object.prototype.hasOwnProperty.call(exports, '__proto__') &&
    Object.defineProperty(exports, '__proto__', {
        enumerable: true,
        value: errors['__proto__']
    });

Object.keys(errors).forEach(function (k) {
    if (k !== 'default' && !Object.prototype.hasOwnProperty.call(exports, k)) exports[k] = errors[k];
});
